overlord/devicestate/devicestate.go

```diff
@@ -475,52 +476,104 @@ func (ms *modelSnapsForRemodel) canHaveUC18PinnedTrack() bool {
 }
 
 type remodelVariant struct {
-	localSnapsRequired bool
+	offline        bool
+	localSnaps     []*snap.SideInfo
+	localSnapPaths []string
 }
 
 type pathSideInfo struct {
 	localSi *snap.SideInfo
 	path    string
 }
 
-func (ro *remodelVariant) UpdateWithDeviceContext(st *state.State,
-	pathSI *pathSideInfo, snapName string, opts *snapstate.RevisionOptions,
+func (ro *remodelVariant) UpdateWithDeviceContext(st *state.State, snapName string, snapID string, opts *snapstate.RevisionOptions,
 	userID int, snapStateFlags snapstate.Flags, tracker snapstate.PrereqTracker,
 	deviceCtx snapstate.DeviceContext, fromChange string) (*state.TaskSet, error) {
 	logger.Debugf("snap %s track changed", snapName)
 	if opts == nil {
 		opts = &snapstate.RevisionOptions{}
 	}
-	if ro.localSnapsRequired {
-		if pathSI == nil {
-			return nil, fmt.Errorf("no snap file provided for %q (track changed)", snapName)
+	if ro.offline {
+		pathSI := ro.sideInfoAndPathFromID(snapID)
+		if pathSI != nil {
+			return snapstateUpdatePathWithDeviceContext(st, pathSI.localSi, pathSI.path, snapName, opts,
+				userID, snapStateFlags, tracker, deviceCtx, fromChange)
 		}
-		return snapstateUpdatePathWithDeviceContext(st, pathSI.localSi, pathSI.path, snapName, opts,
-			userID, snapStateFlags, tracker, deviceCtx, fromChange)
+
+		// TODO: this should also take into account other revisions that we
+		// might have on the system (the revisions in SnapState.Sequence)
+		info, err := snapstate.CurrentInfo(st, snapName)
+		if err != nil {
+			// this case is unexpected, since UpdateWithDeviceContext should
+			// only be called if the snap is already installed
+			if errors.Is(err, &snap.NotInstalledError{}) {
+				return nil, fmt.Errorf("internal error: no snap file provided for %q", snapName)
+			}
+			return nil, err
+		}
+
+		if opts != nil && !opts.Revision.Unset() && info.Revision != opts.Revision {
+			return nil, fmt.Errorf("installed snap %q does not match revision required to be used for offline remodel: %s != %s", snapName, opts.Revision, info.Revision)
+		}
+
+		// this would only occur from programmer error, since
+		// UpdateWithDeviceContext should only be called if either the snap
+		// revision or channel needs to change. once we get here, we know that
+		// the revision is the same, so the channel should be different.
+		if opts == nil || opts.Channel == info.Channel {
+			return nil, fmt.Errorf("internal error: installed snap %q already on channel %q", snapName, info.Channel)
```

Comment:
Is this path tested?

Reply:
No, not tested. But it wouldn't be possible to, since we never call `UpdateWithDeviceContext` in cases where neither the revision or channel has changed. I explained in the comment that this error condition would just be from programmer error.